File: rust/src/api/integration.rs
use super::PostOptions;
use crate::{apis::integration_api, error::Result, models::*, Configuration};

#[derive(Default)]
pub struct IntegrationListOptions {
    pub iterator: Option<String>,
    pub limit: Option<i32>,
    pub order: Option<Ordering>,
}

pub struct Integration<'a> {
    cfg: &'a Configuration,
}

impl<'a> Integration<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self { cfg }
    }

    pub async fn list(
        &self,
        app_id: String,
        options: Option<IntegrationListOptions>,
    ) -> Result<ListResponseIntegrationOut> {
        let IntegrationListOptions {
            iterator,
            limit,
            order,
        } = options.unwrap_or_default();
        integration_api::v1_period_integration_period_list(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodListParams {
                app_id,
                iterator,
                limit,
                order,
            },
        )
        .await
    }

    pub async fn create(
        &self,
        app_id: String,
        integration_in: IntegrationIn,
        options: Option<PostOptions>,
    ) -> Result<IntegrationOut> {
        let PostOptions { idempotency_key } = options.unwrap_or_default();
        integration_api::v1_period_integration_period_create(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodCreateParams {
                app_id,
                integration_in,
                idempotency_key,
            },
        )
        .await
    }

    pub async fn get(&self, app_id: String, integ_id: String) -> Result<IntegrationOut> {
        integration_api::v1_period_integration_period_get(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodGetParams { app_id, integ_id },
        )
        .await
    }

    pub async fn update(
        &self,
        app_id: String,
        integ_id: String,
        integration_update: IntegrationUpdate,
    ) -> Result<IntegrationOut> {
        integration_api::v1_period_integration_period_update(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodUpdateParams {
                app_id,
                integ_id,
                integration_update,
            },
        )
        .await
    }

    pub async fn delete(&self, app_id: String, integ_id: String) -> Result<()> {
        integration_api::v1_period_integration_period_delete(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodDeleteParams { app_id, integ_id },
        )
        .await
    }

    pub async fn get_key(&self, app_id: String, integ_id: String) -> Result<IntegrationKeyOut> {
        integration_api::v1_period_integration_period_get_key(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodGetKeyParams { app_id, integ_id },
        )
        .await
    }

    pub async fn rotate_key(&self, app_id: String, integ_id: String) -> Result<IntegrationKeyOut> {
        integration_api::v1_period_integration_period_rotate_key(
            self.cfg,
            integration_api::V1PeriodIntegrationPeriodRotateKeyParams {
                app_id,
                integ_id,
                idempotency_key: None,
            },
        )
        .await
    }
}

File: javascript/src/api/integration.ts
// this file is @generated
import {
  IntegrationIn,
  IntegrationKeyOut,
  IntegrationOut,
  IntegrationUpdate,
  ListResponseIntegrationOut,
  Ordering,
} from "../openapi";
import { HttpMethod, SvixRequest, SvixRequestContext } from "../request";
import { PostOptions } from "../util";

export interface IntegrationListOptions {
  /** Limit the number of returned items */
  limit?: number;
  /** The iterator returned from a prior invocation */
  iterator?: string | null;
  /** The sorting order of the returned items */
  order?: Ordering;
}

export class Integration {
  public constructor(private readonly requestCtx: SvixRequestContext) {}

  /** List the application's integrations. */
  public list(
    appId: string,
    options?: IntegrationListOptions
  ): Promise<ListResponseIntegrationOut> {
    const request = new SvixRequest(HttpMethod.GET, "/api/v1/app/{app_id}/integration");

    request.setPathParam("app_id", appId);
    request.setQueryParam("limit", options?.limit);
    request.setQueryParam("iterator", options?.iterator);
    request.setQueryParam("order", options?.order);

    return request.send(this.requestCtx);
  }

  /** Create an integration. */
  public create(
    appId: string,
    integrationIn: IntegrationIn,
    options?: PostOptions
  ): Promise<IntegrationOut> {
    const request = new SvixRequest(HttpMethod.POST, "/api/v1/app/{app_id}/integration");

    request.setPathParam("app_id", appId);
    request.setHeaderParam("idempotency-key", options?.idempotencyKey);
    request.body = integrationIn;

    return request.send(this.requestCtx);
  }

  /** Get an integration. */
  public get(appId: string, integId: string): Promise<IntegrationOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/app/{app_id}/integration/{integ_id}"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("integ_id", integId);

    return request.send(this.requestCtx);
  }

  /** Update an integration. */
  public update(
    appId: string,
    integId: string,
    integrationUpdate: IntegrationUpdate
  ): Promise<IntegrationOut> {
    const request = new SvixRequest(
      HttpMethod.PUT,
      "/api/v1/app/{app_id}/integration/{integ_id}"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("integ_id", integId);
    request.body = integrationUpdate;

    return request.send(this.requestCtx);
  }

  /** Delete an integration. */
  public delete(appId: string, integId: string): Promise<void> {
    const request = new SvixRequest(
      HttpMethod.DELETE,
      "/api/v1/app/{app_id}/integration/{integ_id}"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("integ_id", integId);

    return request.sendNoResponseBody(this.requestCtx);
  }

  /**
   * Get an integration's key.
   *
   * @deprecated
   */
  public getKey(appId: string, integId: string): Promise<IntegrationKeyOut> {
    const request = new SvixRequest(
      HttpMethod.GET,
      "/api/v1/app/{app_id}/integration/{integ_id}/key"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("integ_id", integId);

    return request.send(this.requestCtx);
  }

  /** Rotate the integration's key. The previous key will be immediately revoked. */
  public rotateKey(
    appId: string,
    integId: string,
    options?: PostOptions
  ): Promise<IntegrationKeyOut> {
    const request = new SvixRequest(
      HttpMethod.POST,
      "/api/v1/app/{app_id}/integration/{integ_id}/key/rotate"
    );

    request.setPathParam("app_id", appId);
    request.setPathParam("integ_id", integId);
    request.setHeaderParam("idempotency-key", options?.idempotencyKey);

    return request.send(this.requestCtx);
  }
}
